File: lombok/src/main/java/cheese/yun/lombok/Coupon.java
package cheese.yun.lombok;

import lombok.Data;
import lombok.Setter;

import javax.persistence.*;

@Entity
@Table(name = "coupon")
@Data
@Setter
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @ManyToOne
    private Member member;

    public Coupon(Member member) {
        this.member = member;
    }
}

File: jpa-n-plus-1/src/main/kotlin/com/example/jpanplus1/copon/Coupon.kt
package com.example.jpanplus1.copon

import com.example.jpanplus1.member.Member
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.LocalDateTime
import javax.persistence.*

@Entity
@Table(name = "coupon")
class Coupon private constructor() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0

    @Column(name = "code", nullable = false, unique = true)
    lateinit var code: String
        private set

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    lateinit var createdAt: LocalDateTime
        private set

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    lateinit var updatedAt: LocalDateTime
        private set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = true, updatable = false)
    lateinit var member: Member
        private set

    constructor(code: String) : this() {
        this.code = code
    }

}